hail/src/main/scala/is/hail/io/fs/AzureStorageFS.scala

```diff
@@ -0,0 +1,388 @@
+    new WrappedPositionedDataOutputStream(os)
+  }
+
+  def delete(filename: String, recursive: Boolean): Unit = {
+    if (this.exists(filename)) {
+      val (account, container, path) = getAccountContainerPath(filename)
+      val blobClient: BlobClient = this.getBlobClient(filename)
+
+      val fileStatus = this.fileStatus(filename)
+
+      if (recursive) {
+        val blobContainerClient = this.getContainerClient(filename)
+
+        val directoryContents = blobContainerClient.listBlobsByHierarchy(dropTrailingSlash(path) + "/")
+        var seenBlobItems: Set[String] = Set()
+        val iter = directoryContents.iterator()
+
+        def deleteRecursively(): Unit = {
+          while (iter.hasNext) {
+            val blobItem: BlobItem = iter.next
+            if (seenBlobItems.contains(blobItem.getName)) return
+            seenBlobItems += blobItem.getName
+            val blobFileName = s"hail-az://$account/$container/${blobItem.getName}"
+            // recursively delete all files/subfolders
+            this.delete(blobFileName, recursive)
+          }
+        }
+        deleteRecursively()
+
+        if (fileStatus.isFile && blobClient.exists()) blobClient.delete()
+      }
+      else {
+          if (fileStatus.isDirectory) {
+            throw new IllegalArgumentException("Cannot non-recursively delete a directory")
+          }
+          if (blobClient.exists()) blobClient.delete()
+      }
+    }
+  }
+
+  def listStatus(filename: String): Array[FileStatus] = {
+    val (account, container, path) = getAccountContainerPath(filename)
+    val blobContainerClient: BlobContainerClient = getContainerClient(filename)
+
+    val statusList: mutable.MutableList[FileStatus] = mutable.MutableList()
+    var blobNameSet: Set[String] = Set()
+
+    var seenBlobItems: Set[String] = Set()
+
+    val prefix = dropTrailingSlash(path) + "/"
+    val directoryContents = blobContainerClient.listBlobsByHierarchy(prefix)
+    val iter = directoryContents.iterator()
+
+    def populateStatList(): Unit = {
+      while (iter.hasNext) {
+        val blobItem: BlobItem = iter.next
+
+        if (seenBlobItems.contains(blobItem.getName)) return
+        seenBlobItems += blobItem.getName
+
+        // find the first slash after the prefix
+        val indexOfFirstSlashAfterPrefix: Int = blobItem.getName.indexOf("/", prefix.length)
+        var blobName = blobItem.getName
+        // if there is a slash after the prefix, then this blob is under a subdirectory
+        if (indexOfFirstSlashAfterPrefix != -1) {
+          blobName = blobItem.getName.substring(0, indexOfFirstSlashAfterPrefix)
+        }
+
+        if (!blobNameSet.contains(blobName)) {
+          blobNameSet = blobNameSet + blobName
+          val blobFileName = s"hail-az://$account/$container/$blobName"
+          statusList += fileStatus(blobFileName)
+        }
+      }
+    }
+    populateStatList()
+
+    statusList.toArray
+  }
+
+  def glob(filename: String): Array[FileStatus] = {
```

Comment:
This is pretty much the same as the Google implementation, they just use different implementations of `listStatus`. Let's abstract these. Maybe just put them on the trait and have hadoop override?

Reply:
Actually the way the filenames are deconstructed and the prefix is constructed are also different, so I don't know if this abstraction makes sense (it also looks like only 2 methods are scheme-specific)

Reply:
I think we can reuse this if we change the signature slightly. What does glob actually need of the filename here? It looks like it needs a prefix, so the bucket or account/container pair, and it needs the path because it slits it into `components`. So what if we make a helper with the following signature `glob(prefix: String, path: String)` and the Google/Azure implementations call that after splitting it appropriately. It would also avoid this weird re-assembling of the string on line 286 happening inside the glob implementation.